src/commonMain/kotlin/fr/acinq/lightning/wire/LiquidityAds.kt

```diff
@@ -20,140 +24,352 @@ import fr.acinq.lightning.utils.sat
+            return fundingRates
+                .filterIsInstance<FundingLease.Basic>()
+                .firstOrNull { it.minAmount <= requestedAmount && requestedAmount <= it.maxAmount }
         }
 
         fun write(out: Output) {
-            LightningCodecs.writeU16(leaseDuration, out)
-            LightningCodecs.writeU16(fundingWeight, out)
-            LightningCodecs.writeU16(leaseFeeProportional, out)
-            LightningCodecs.writeU32(leaseFeeBase.sat.toInt(), out)
-            LightningCodecs.writeU16(maxRelayFeeProportional, out)
-            LightningCodecs.writeU32(maxRelayFeeBase.msat.toInt(), out)
+            LightningCodecs.writeU16(fundingRates.size, out)
+            fundingRates.forEach { it.write(out) }
+            val encoded = PaymentType.encode(paymentTypes)
+            LightningCodecs.writeU16(encoded.size, out)
+            LightningCodecs.writeBytes(encoded, out)
         }
 
         companion object {
-            fun read(input: Input): LeaseRate = LeaseRate(
-                leaseDuration = LightningCodecs.u16(input),
-                fundingWeight = LightningCodecs.u16(input),
-                leaseFeeProportional = LightningCodecs.u16(input),
-                leaseFeeBase = LightningCodecs.u32(input).sat,
-                maxRelayFeeProportional = LightningCodecs.u16(input),
-                maxRelayFeeBase = LightningCodecs.u32(input).msat,
+            fun read(input: Input): WillFundRates {
+                val fundingRatesCount = LightningCodecs.u16(input)
+                val fundingRates = (0 until fundingRatesCount).mapNotNull { FundingLease.read(input) }
+                val paymentTypes = PaymentType.decode(LightningCodecs.bytes(input, LightningCodecs.u16(input)))
+                return WillFundRates(fundingRates, paymentTypes)
+            }
+        }
+    }
+
+    /** Provide inbound liquidity to a remote peer that wants to purchase a liquidity ads. */
+    data class WillFund(val leaseWitness: FundingLeaseWitness, val signature: ByteVector64) {
+        fun write(out: Output) {
+            leaseWitness.write(out)
+            LightningCodecs.writeBytes(signature, out)
+        }
+
+        companion object {
+            fun read(input: Input): WillFund = WillFund(
+                leaseWitness = FundingLeaseWitness.read(input),
+                signature = LightningCodecs.bytes(input, 64).byteVector64(),
             )
         }
     }
 
     /** Request inbound liquidity from a remote peer that supports liquidity ads. */
-    data class RequestRemoteFunding(val fundingAmount: Satoshi, val leaseStart: Int, val rate: LeaseRate) {
-        private val leaseExpiry: Int = leaseStart + rate.leaseDuration
-        val requestFunds: ChannelTlv.RequestFunds = ChannelTlv.RequestFunds(fundingAmount, rate.leaseDuration, leaseExpiry)
+    data class RequestFunds(val requestedAmount: Satoshi, val fundingLease: FundingLease, val paymentDetails: PaymentDetails) {
```

Comment:
Otherwise it sounds like we are creating an invoice:

```suggestion
    data class RequestFunding(val requestedAmount: Satoshi, val fundingLease: FundingLease, val paymentDetails: PaymentDetails) {
```

Reply:
Good point, done.